doc/csharp/csharp-null-ref-types-proposal.md

```diff
@@ -0,0 +1,170 @@
+class may cause errors (not warnings) in user code that has already been derived 
+from the base class:
+
+> *error CS8765: Nullability of type of parameter 'request' doesn't match 
+overridden member (possibly because of nullability attributes).*
+
+Note: the advantages of adding these attributes to the service methods is 
+questionable as the only caller of these methods is the server hosting the 
+services, not user code. Therefore this may not be an issue if it is not needed 
+to be done.
+
+## Disadvantages of this propsal
+
+There are a few places where adding annotations or attributes to the generated
+code to help with the static flow analysis could be useful. These are:
+- Checking non-null setters for certain fields in a message at compile time,
+ such as fields that are strings. Runtime checks are still done in the
+ generated code.
+- Checking a non-null message or stream is passed to a gRPC service call.
+- Specifying a non-null message or stream is returned from a gRPC service call.
+
+By not supporting nullable reference types in the generated code these compile
+time checks cannot be done.
+
+## Outside the scope of this proposal
+
+### Changes to related packages
+
+Changes to `Google.Protobuf` library and other libraries 
+(such as `Google.Api.CommonProtos`) are outside the scope of this proposal.
+
+### Exposing protocol buffers optional fields as nullable properties
+
+See https://github.com/protocolbuffers/protobuf/issues/9083
+
+This has been rejected by the protocol buffers team.
+
+> "We get requests in some other languages that have null support to allow 
+optional fields to have nullable properties. The main problem is that we like to 
+keep languages consistent and other languages don't have those same options. 
+Additionally, it's not clear with some primitive fields what a null value would 
+be (is it the default? What if users meant to set it to that option?). For those 
+reasons, we will not support this in the near future.”
+
+## Alternative options to this proposal
+
+If in the future we want to support nullable reference types in the generated 
+code then this should be an *opt-in* feature so as to not break existing 
+projects.
+
+We need to consider:
+- Options for enabling this feature
+- What code changes need to be made in the generated code
+
+Options for enabling this feature:
+
+- Generate conditional code and the user has to define a constant in the project files to enable it
+  - The generated conditional code might be messy
+  - Could change `Grpc.Tools` to define the constant if it detects project has nullable enabled
+
+- Generate code that relies on definitions of nullable attributes for older framework versions
+  - Like those in https://github.com/grpc/grpc-dotnet/blob/master/src/Shared/NullableAttributes.cs
+  - Where would these definitions of nullable attributes be defined?
+  Don’t want to define them in every generated C# file. Is there a single 
```

Comment:
Putting the annotations in any shared library would mean they'd have to be public rather than internal, which then causes issues with potential collisions IME.